src/docs/sphinx/dev_guide/weak_form.rst

```diff
@@ -0,0 +1,261 @@
+vector of weighted residuals, using H1 test functions.
+
+Now that we have this ``WeakForm`` object created, we can use the
+following functions to define integral terms (depending on their
+dimensionality):
+
+1. Integrals of the form:
+   :math:`\displaystyle \iint_\Omega \psi \cdot f_0 + \nabla \psi : f_1 \; da`
+
+   .. code-block:: cpp
+
+      WeakForm::AddAreaIntegral([](auto ... args){
+      	auto f0 = ...;
+      	auto f1 = ...;
+      	return std::tuple{f0, f1};
+      }, domain_of_integration);
+
+2. Integrals of the form:
+   :math:`\displaystyle \iiint_\Omega \psi \cdot f_0 + \nabla \psi : f_1 \; dv`
+
+   .. code-block:: cpp
+
+      WeakForm::AddVolumeIntegral([](auto ... args){
+      	auto f0 = ...;
+      	auto f1 = ...;
+      	return std::tuple{f0, f1};
+      }, domain_of_integration);
+
+3. Integrals of the form:
+   :math:`\displaystyle \iint_{\partial \Omega} \psi \cdot f \; da`
+
+   .. code-block:: cpp
+
+      WeakForm::AddSurfaceIntegral([](auto ... args){
+      	auto f0 = ...;
+      	return f0;
+      }, domain_of_integration);	
+
+So, for our problem (since we assumed 3D earlier) we can make an
+``Add****Integral()`` call for each of the integral terms in the
+original residual. In each of these functions, the first argument is the
+integrand (a lambda function or functor returning :math:`\{f_0, f_1\}`),
+and the second argument is the domain of integration. Let's start with
+the stress response term:
+
+.. code-block:: cpp
+
+   // The integrand lambda function is passed the spatial position of the quadrature point,
+   // as well as a {value, derivative} tuple for the trial space.
+   residual.AddVolumeIntegral([](auto x, auto disp){
+     
+     // Here, we unpack the {value, derivative} tuple into separate variables
+     auto [u, grad_u] = disp;
+     
+     // WeakForm expects us to return a tuple of the form {f_0, f_1} (see table above)
+     auto f0 = zero{}; // for this term, the f_0 term is identically zero.
+     auto f1 = stress(grad_u); // call some constitutive model for the material in this domain
+     
+     return std::tuple{f0, f1};
+     
+   }, mesh);
+
+The other terms follow a similar pattern. For the body force:
+
+.. code-block:: cpp
+
+   residual.AddVolumeIntegral([](auto x, auto disp /* unused */){
+     
+     // WeakForm::AddVolumeIntegral() expects us to return a tuple of the form {f_0, f_1}
+     auto f0 = b(x); // evaluate the body-force at the location of the quadrature point
+     auto f1 = zero{}; // for this term, the f_1 term is identically zero
+     
+     return std::tuple{f0, f1};
+     
+   }, mesh);
+
+And finally, for the surface tractions:
+
+.. code-block:: cpp
+
+   // WeakForm::AddSurfaceIntegral() only expects us to return f, so we don't need a tuple
+   residual.AddSurfaceIntegral([](auto x, auto disp /* unused */){
+     return traction(x); // evaluate the traction at the location of the quadrature point
+   }, surface_mesh);
```

Comment:
For now it takes a mesh and integrating over the all of the boundary faces, so the code snippet is consistent with the current implementation. Although, I agree that this isn't what we actually want. I'm tempted to leave this comment as-is, since it is at least consistent, and then update it when we change the interface to the one that supports submeshes or attribute sets.